crates/matrix-sdk/src/widget/matrix.rs

```diff
@@ -309,38 +318,137 @@ impl MatrixDriver {
                 | "m.room.encrypted"
-        );
-
-        if filtered {
-            trace!("To-device message of type <{event_type}> filtered out by widget driver.",);
-        }
-        filtered
+        )
     }
 
-    /// It will ignore all devices where errors occurred or where the device is
-    /// not verified or where th user has a has_verification_violation.
+    /// If the room the widget is in is encrypted, then the to-device message
+    /// will be encrypted. If one of the named devices does not exist, then
+    /// the call will fail with an error.
     pub(crate) async fn send_to_device(
         &self,
         event_type: ToDeviceEventType,
-        encrypted: bool,
         messages: BTreeMap<
             OwnedUserId,
             BTreeMap<DeviceIdOrAllDevices, Raw<AnyToDeviceEventContent>>,
         >,
-    ) -> Result<send_event_to_device::v3::Response> {
+    ) -> Result<SendToDeviceEventResponse> {
+        // TODO: block this at the negotiation stage, no reason to let widget believe
+        // they can do that
+        if Self::is_internal_type(&event_type.to_string()) {
+            warn!("Widget tried to send internal to-device message <{}>, ignoring", event_type);
+            // Silently return a success response, the widget will not receive the message
+            // TODO! this will later be blocked at negotiation level
+            return Ok(Default::default());
+        }
+
         let client = self.room.client();
 
-        let request = if encrypted {
-            return Err(Error::UnknownError(
-                "Sending encrypted to-device events is not supported by the widget driver.".into(),
-            ));
-        } else {
-            RumaToDeviceRequest::new_raw(event_type, TransactionId::new(), messages)
-        };
+        let mut failures: BTreeMap<OwnedUserId, Vec<OwnedDeviceId>> = BTreeMap::new();
 
-        let response = client.send(request).await;
+        let room_encrypted = self
+            .room
+            .latest_encryption_state()
+            .await
+            .map(|s| s.is_encrypted())
+            // Default consider encrypted
+            .unwrap_or(true);
+
+        if room_encrypted {
+            trace!("Sending to-device message in encrypted room <{}>", self.room.room_id());
+
+            //  Attempt to re-batch these by content to batch them up into a single request
+            let mut content_grouped: BTreeMap<
+                &str,
+                BTreeMap<OwnedUserId, Vec<DeviceIdOrAllDevices>>,
+            > = BTreeMap::new();
+
+            for (user_id, device_map) in messages.iter() {
+                for (device_id, content) in device_map.iter() {
+                    content_grouped
+                        .entry(content.json().get())
+                        .or_default()
+                        .entry(user_id.clone())
+                        .or_default()
+                        .push(device_id.to_owned());
+                }
+            }
```

Comment:
This one is a bit trippy, you are reversing the mapping from user -> device -> content to content -> user -> device.

As far as I understand this is an optimization so we can use the function that encrypts the same content for multiple devices.

Could you document this fact, it isn't immediately obvious why this is needed.

Renaming the variable to `content_to_recipients_map` or some similar more obvious name might help as well.

Reply:
done